dataverse-webapp/src/main/java/edu/harvard/iq/dataverse/interceptors/RestrictedInterceptor.java

```diff
@@ -0,0 +1,146 @@
+import javax.interceptor.Interceptor;
+import javax.interceptor.InvocationContext;
+import java.lang.reflect.Method;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.EnumSet;
+import java.util.List;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import static edu.harvard.iq.dataverse.interceptors.InterceptorCommons.createName;
+
+@Interceptor
+@Restricted
+public class RestrictedInterceptor {
+
+    private PermissionDataProcessor permissionDataProcessor = new PermissionDataProcessor();
+
+    @EJB
+    private PermissionServiceBean permissionService;
+
+    @Inject
+    private DataverseRequestServiceBean requestService;
+
+    //-------------------- LOGIC --------------------
+
+    @AroundInvoke
+    public Object callRestricted(InvocationContext ctx) throws Exception {
+        DataverseRequest request = requestService.getDataverseRequest();
+        Method method = ctx.getMethod();
+        Set<RestrictedObject> restrictedObjects = permissionDataProcessor.gatherPermissionRequirements(method, ctx.getParameters());
+        if (restrictedObjects.isEmpty()) {
+            throw new RuntimeException("Service method: " + createName(method) + " does not define required permissions.");
+        }
+
+        logBasicInfo(fetchOrCreateLogRecord(ctx, method, request), restrictedObjects);
+        List<MissingPermissions> missingPermissions = checkPermissions(restrictedObjects, request);
+        if (!missingPermissions.isEmpty()) {
+            throw createMissingPermissionException(missingPermissions, method, request);
+        }
+
+        try {
+            return ctx.proceed();
+        } catch (EJBException ee) {
+            throw new CommandException("Service method: " + createName(method) + " failed: " + ee.getMessage(),
+                    ee.getCausedByException(), null);
+        }
+    }
+
+    private PermissionException createMissingPermissionException(
+            List<MissingPermissions> missingPermissions, Method method, DataverseRequest request) {
+        MissingPermissions firstMissing = missingPermissions.get(0);
+        return new PermissionException("Can't execute method " + method
+                + ", because request: " + request
+                + " is missing permissions: " + firstMissing.missing
+                + " on Object: [" + firstMissing.dvObjectName
+                + "]. All missing permissions: " + missingPermissions,
+                null, firstMissing.required, firstMissing.dvObject);
+    }
+
+    // -------------------- PRIVATE ---------------------
+
+    private ActionLogRecord fetchOrCreateLogRecord(InvocationContext ctx, Method method, DataverseRequest request) {
+        return (ActionLogRecord) ctx.getContextData().computeIfAbsent(InterceptorCommons.LOG_RECORD_KEY,
+                k -> new ActionLogRecord(ActionType.Command, createName(method))
+                        .setUserIdentifier(request.getUser().getIdentifier()));
+    }
```

Comment:
Yes. That was a kind of defensive behavior in case of LoggedCallInterceptor issues, but in fact this would not be persisted at the end, so I can remove it completely.